lambda_services/tests/apbsjob.py

```diff
@@ -0,0 +1,99 @@
+        """The job consists of input and output files from an APBS job
+
+        Args:
+            jobid (str): The unique string for an Azure blog or AWS bucket
+            file_path (str): The absolute path to the mounted directory
+            file_list (List): The list of files for the job
+
+        Raises:
+            TypeError: The file_path is not a directory
+        """
+        # NOTE: The file list should be something like:
+        # apbs_end_time
+        # apbs_exec_exit_code.txt
+        # apbs_input_files
+        # apbs_output_files
+        # apbs_start_time
+        # apbs_status
+        # apbs_stderr.txt
+        # apbs_stdout.txt
+        # apbsinput.in
+        # *.in
+        # *.prq
+
+        self.job_type = JOBTYPE.APBS.name.lower()
+        self._logger = getLogger(__class__.__name__)
+        super().__init__(jobid, file_path, file_list)
+
+    def get_memory_usage(self):
+        """Get the memory used and high water memory that could have been used
+
+        Returns:
+            Dict: Total & high memory usage found in apbs_stdout.txt file
+        """
+        # NOTE: This could be done by looking at the dime values in
+        #       the apbsinput.in file and taking the x, y, and z values
+        #       and multiplying them times 0.00019073486 and then if
+        #       the elec type is mg-auto you could multiply that by 2.
+        #
+        #       Otherwise, you can parse the "Final memory usage" from
+        #       the apbs_stdout.txt file. The line looks like:
+        # Final memory usage:  0.001 MB total, 2666.345 MB high water
+        mem_used = {"total": None, "high": None}
+        lines = get_contents(self.file_list[f"{self.job_type}_stdout.txt"])
+        for line in lines:
+            if line.startswith("Final memory usage"):
+                self._logger.debug("MEM LINE: %s", line)
+                values = re.findall("\d+\.?\d+", line)  # noqa W605
+                self._logger.debug("VALUES: %s", values)
+                mem_used["total"] = values[0]
+                mem_used["high"] = values[1]
+        return mem_used
+
+    def build_job_file(self):
+        """
+        Create an apbs-job.json file to simulate what the Web/Gui/React
+        application would create if a user submitted the job.
+        """
+        job = {
+            "form": {
+                "job_id": self.job_id,
+                "invoke_method": "v2",
+                "file_list": [],
+            }
+        }
+        # Open up the apbs_input_files to find the files needed
+        # to build a apbs-job.json that looks like the following:
+        # {
+        #   "job_id": "000buomdp2",
+        #   "file_list": [
+        #       "apbsinput.in",
+        #       "000buomdp2.pqr"
+        #   ],
+        # }
+
+        for filename in self.file_list:
+            if filename.endswith(".pqr") or filename in "apbsinput.in":
+                job["form"]["file_list"].append(filename)
```

Comment:
Did this end up executing a job for you? Asking because the expected format for an APBS job would look like the following, ...
```
{
    "form": {
        "filename": "sample.in"
    }
}
```

...where name of the input file would be searched on the key "filename".  The example format in the comments looks like that which the API Gateway uses to create upload URLs

Reply:
I would have thought that would be the case but...

There are only 10 out of 1387 APBS jobs that don't have `apbsinput.in` in the file list:
```
grep '.in' cache_meta/cache_apbs.txt  | grep -v 'apbsinput.in' 
2x0qdp4heq/ ['apbs_end_time', 'apbs_exec_exit_code.txt', 'apbs_input_files', 'apbs_output_files', 'apbs_start_time', 'apbs_status', 'apbs_stdout.txt', 'xlq7123yfe.cube', 'xlq7123yfe.cube.gz', 'xlq7123yfe.in', 'xlq7123yfe.pqr', 'xlq7123yfe.pqr.dx.gz']
m9deo8dd9d/ ['apbs_end_time', 'apbs_exec_exit_code.txt', 'apbs_input_files', 'apbs_output_files', 'apbs_start_time', 'apbs_status', 'apbs_stdout.txt', 'emzx513rdn.cube', 'emzx513rdn.cube.gz', 'emzx513rdn.in', 'emzx513rdn.pqr', 'emzx513rdn.pqr.dx', 'emzx513rdn.pqr.dx.gz']
nh00q41bci/ ['2c5wt7ds1n.cube', '2c5wt7ds1n.in', '2c5wt7ds1n.pqr', '2c5wt7ds1n.pqr.dx', 'apbs_input_files', 'apbs_output_files', 'apbs_start_time', 'apbs_status', 'apbs_stdout.txt']
nspixokx7a/ ['TP301S-0.in', 'TP301S-0.pqr', 'TP301S-0.pqr.dx', 'apbs_input_files', 'apbs_start_time', 'apbs_status']
t0lyg766n5/ ['apbs_input_files', 'apbs_start_time', 'apbs_status', 'apbs_stdout.txt', 'sphere-ligand-25-pot.dx', 'sphere_R25A.in', 'sphere_R25A_sort.pqr']
tznboiu8o7/ ['PB_9NZ4RDHS.dx', 'PB_9NZ4RDHS.in', 'PB_9NZ4RDHS.pqr', 'ttt.pdb']
wx6yb38of6/ ['TauWT.cube', 'TauWT.in', 'TauWT.pqr', 'TauWT.pqr.dx.gz', 'apbs_end_time', 'apbs_exec_exit_code.txt', 'apbs_input_files', 'apbs_output_files', 'apbs_start_time', 'apbs_status', 'apbs_stdout.txt']
30ftimra41/ ['apbs_end_time', 'apbs_exec_exit_code.txt', 'apbs_input_files', 'apbs_output_files', 'apbs_start_time', 'apbs_status', 'apbs_stdout.txt', 'n4e2leqtig.in', 'n4e2leqtig.pqr', 'n4e2leqtig.pqr.dx.gz']
6v238et3l3/ ['io.mc', 'potential3.dx', 'working.in', 'working.pdb']
qhrdzp8lui/ ['apbs.in', 'io.mc', 'run.pqr', 'sspl.dx']
```

If you can find a pattern in the 10 jobs above, I would be glad to implement it. Some of them like `qhrdzp8lui/` does not look like it ran.

Reply:
It's hard to guess what happened to those last two, but with the others, uploading whatever's in `apbs_input_files` *should* provide the files needed to run a job since the *.in file would be among those?  Then specifying said *.in file within the JSON form like in my comment above should tell it to execute with that file.

So perhaps an approach could be to assume `apbsinput.in` since that appears to be the common method, but if it doesn't exist then use a *.in file among those found in the `job_id` directory.

Does that address what you were alluding to?

Reply:
@Eo300 The latest update to apbsjob.py addresses your comment. Thanks for the clarification.